mistral/bitstream.cc

```diff
@@ -29,108 +29,13 @@ struct MistralBitgen
-        cv->opt_b_set(CycloneV::OCT_DONE_DIS, true);
-        cv->opt_r_set(CycloneV::OPT_A, 0x1dff);
-        if (!ctx->setting<bool>("compress_rbf", false))
             cv->opt_r_set(CycloneV::OPT_B, 0xffffff40adffffffULL);
-        else
+        } else
             cv->opt_r_set(CycloneV::OPT_B, 0xffffff402dffffffULL);
-        cv->opt_b_set(CycloneV::RELEASE_CLEARS_BEFORE_TRISTATES_DIS, true);
-        cv->opt_b_set(CycloneV::RETRY_CONFIG_ON_ERROR_EN, true);
-        cv->opt_r_set(CycloneV::START_UP_CLOCK, 0x3F);
-        // Default inversion
-        write_default_inv();
-    }
-
-    void write_default_inv()
-    {
-        // Some PNODEs are inverted by default. Set them up here.
-        for (const auto &pn2r : cv->get_all_p2r()) {
-            const auto &pn = pn2r.first;
-            auto pt = CycloneV::pn2pt(pn);
-            auto pi = CycloneV::pn2pi(pn);
-
-            switch (CycloneV::pn2bt(pn)) {
-            case CycloneV::HMC: {
-                // HMC OE are inverted to set OE=0, i.e. unused pins floating
-                // TODO: handle the case when we are using the HMC or HMC bypass
-                std::string name(CycloneV::port_type_names[pt]);
-                if (name.compare(0, 5, "IOINT") != 0 || name.compare(name.size() - 2, 2, "OE") != 0)
-                    continue;
-                cv->inv_set(pn2r.second, true);
-                break;
-            };
-            // HPS IO - TODO: what about when we actually support the HPS primitives?
-            case CycloneV::HPS_BOOT: {
-                switch (pt) {
-                case CycloneV::CSEL_EN:
-                case CycloneV::BSEL_EN:
-                case CycloneV::BOOT_FROM_FPGA_READY:
-                case CycloneV::BOOT_FROM_FPGA_ON_FAILURE:
-                    cv->inv_set(pn2r.second, true);
-                    break;
-                case CycloneV::CSEL:
-                    if (pi < 2)
-                        cv->inv_set(pn2r.second, true);
-                    break;
-                case CycloneV::BSEL:
-                    if (pi < 3)
-                        cv->inv_set(pn2r.second, true);
-                    break;
-                default:
-                    break;
-                };
-                break;
-            };
-            case CycloneV::HPS_CROSS_TRIGGER: {
-                if (pt == CycloneV::CLK_EN)
-                    cv->inv_set(pn2r.second, true);
-                break;
-            };
-            case CycloneV::HPS_TEST: {
-                if (pt == CycloneV::CFG_DFX_BYPASS_ENABLE)
-                    cv->inv_set(pn2r.second, true);
-                break;
-            };
-            case CycloneV::GPIO: {
-                // Ignore GPIO used by the design
-                BelId bel = ctx->bel_by_block_idx(CycloneV::pn2x(pn), CycloneV::pn2y(pn), id_MISTRAL_IO,
-                                                  CycloneV::pn2bi(pn));
-                if (bel != BelId() && ctx->getBoundBelCell(bel) != nullptr)
```

Comment:
I need to check this PR on hardware, but note that this only sets the IO default inverters when they are unused. Do we need to add code to clear the default inverters when they are used instead?

Reply:
Possibly so yes.  At this point mistral (I think) correctly sets the state for an empry fpga for the given package when calling clear().  So you only need to delta from that.